### PR TITLE
Refactor BlockHeader / Add version in BlockHeaderBase / Modify Block Operator

### PR DESCRIPTION
**Pull request by kaikawaliu**
## Description

Most of the content is for refactoring.
Added version for all the block data structure
Also added prevHash into BlockHeaderBase

Cleaned some logs after the changes also.

### Implementation

- [x] **ready for review**

### Integration Test (Core Team)

~~- [ ] local machine test~~

- [x] small-scale cloud test

### CODE REVIEW

File: constants.xml [hunk 1]
```diff
@@ -9,6 +9,15 @@
         <MAX_ENTRIES_FOR_DIAGNOSTIC_DATA>25</MAX_ENTRIES_FOR_DIAGNOSTIC_DATA>
         <CHAIN_ID>62</CHAIN_ID>
     </general>
+    <version>
+        <MSG_VERSION>1</MSG_VERSION>
```
Comment (ansnunez) [inline constants.xml]: I think you forgot to delete MSG_VERSION under general

File: src/libData/BlockData/BlockHeader/FallbackBlockHeader.cpp [hunk 2]
```diff
@@ -107,19 +105,28 @@ const PubKey& FallbackBlockHeader::GetLeaderPubKey() const {
...
 }
 
 bool FallbackBlockHeader::operator<(const FallbackBlockHeader& header) const {
-  return std::tie(header.m_fallbackEpochNo, header.m_fallbackDSEpochNo,
-                  header.m_fallbackState, header.m_hashset,
-                  header.m_leaderConsensusId, header.m_shardId) >
-         std::tie(m_fallbackEpochNo, m_fallbackDSEpochNo, m_fallbackState,
-                  m_hashset, m_leaderConsensusId, m_shardId);
+  // To compare, first they must be of identical epochno and state
+  if (!(std::tie(m_version, m_fallbackEpochNo, m_fallbackDSEpochNo,
```
Comment (ansnunez) [inline src/libData/BlockData/BlockHeader/FallbackBlockHeader.cpp]: m_version needed?

File: src/libData/BlockData/BlockHeader/FallbackBlockHeader.cpp [hunk 3]
```diff
@@ -107,19 +105,28 @@ const PubKey& FallbackBlockHeader::GetLeaderPubKey() const {
...
+                 m_fallbackState) ==
+        std::tie(m_version, header.m_fallbackEpochNo,
```
Comment (ansnunez) [inline src/libData/BlockData/BlockHeader/FallbackBlockHeader.cpp]: Should be `header.m_version`?

File: src/libData/BlockData/BlockHeader/VCBlockHeader.cpp [hunk 4]
```diff
@@ -99,28 +97,27 @@ const vector<pair<PubKey, Peer>>& VCBlockHeader::GetFaultyLeaders() const {
...
+                   header.m_FaultyLeaders));
 }
 
 bool VCBlockHeader::operator<(const VCBlockHeader& header) const {
   // To compare, first they must be of identical epochno and state
-  return (m_VieWChangeDSEpochNo == header.m_VieWChangeDSEpochNo) &&
-         (m_VieWChangeEpochNo == header.m_VieWChangeEpochNo) &&
-         (m_ViewChangeState == header.m_ViewChangeState) &&
+  return BlockHeaderBase::operator==(header) &&
+         (std::tie(m_VieWChangeDSEpochNo, m_VieWChangeEpochNo,
```
Comment (ansnunez) [inline src/libData/BlockData/BlockHeader/VCBlockHeader.cpp]: Should version be included?

**Comment by ansnunez** [comment]
@kaikawaliu Why are `CheckBlockTypeIsMicro` and `CheckBlockTypeIsFinal` removed?

**Comment by kaikawaliu** [comment]
> @kaikawaliu Why are `CheckBlockTypeIsMicro` and `CheckBlockTypeIsFinal` removed?

Because the type field is no longer used. which is useful when implementing the homemade serialization method, now microblock and finalblock has their own specific data structure and serialization/deserialization function in protobuf.

File: src/libData/BlockData/BlockHeader/VCBlockHeader.cpp (hunk 4 above)
Comment (kaikawaliu) [reply src/libData/BlockData/BlockHeader/VCBlockHeader.cpp]: version must be equal and which is included in the BlockHeaderBase

File: src/libData/BlockData/BlockHeader/FallbackBlockHeader.cpp (hunk 3 above)
Comment (kaikawaliu) [reply src/libData/BlockData/BlockHeader/FallbackBlockHeader.cpp]: oops

File: src/libData/BlockData/BlockHeader/BlockHeaderBase.cpp [hunk 5]
```diff
@@ -38,6 +42,35 @@ BlockHash BlockHeaderBase::GetMyHash() const {
...
+  m_prevHash = prevHash;
+}
+
+bool BlockHeaderBase::operator==(const BlockHeaderBase& header) const {
+  return std::tie(m_version, m_committeeHash, m_prevHash) ==
+         std::tie(header.m_version, header.m_committeeHash, header.m_prevHash);
+}
+
+bool BlockHeaderBase::operator<(const BlockHeaderBase& header) const {
+  return m_version < header.m_version;
```
Comment (bb111189) [inline src/libData/BlockData/BlockHeader/BlockHeaderBase.cpp]: Is this check sufficient?

Comment (kaikawaliu) [reply src/libData/BlockData/BlockHeader/BlockHeaderBase.cpp]: yeah as there is no other usage of this operator currently, should be enough as the hash shouldn't be used for comparison anyway.

**Pull request merged**